subsys/mgmt/mcumgr/smp/src/client.c

```diff
@@ -0,0 +1,295 @@
+		sys_slist_append(&smp_client_data.cmd_free_list, &smp_cmd_req_bufs[i].node);
+	}
+	return 0;
+}
+
+static struct smp_client_cmd_req *smp_client_cmd_req_allocate(void)
+{
+	sys_snode_t *cmd_node;
+	struct smp_client_cmd_req *req;
+
+	cmd_node = sys_slist_get(&smp_client_data.cmd_free_list);
+	if (!cmd_node) {
+		return NULL;
+	}
+
+	req = SYS_SLIST_CONTAINER(cmd_node, req, node);
+
+	return req;
+}
+
+static void smp_cmd_add_to_list(struct smp_client_cmd_req *cmd_req)
+{
+	if (sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		/* Enable timer */
+		k_work_reschedule(&smp_client_data.work_delay, K_MSEC(CONFIG_SMP_CMD_RETRY_TIME));
+	}
+	sys_slist_append(&smp_client_data.cmd_list, &cmd_req->node);
+}
+
+static void smp_client_cmd_req_free(struct smp_client_cmd_req *cmd_req)
+{
+	smp_client_buf_free(cmd_req->nb);
+	cmd_req->nb = NULL;
+	sys_slist_find_and_remove(&smp_client_data.cmd_list, &cmd_req->node);
+	/* Add to free list*/
+	sys_slist_append(&smp_client_data.cmd_free_list, &cmd_req->node);
+
+	if (sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		/* cancel delay*/
+		k_work_cancel_delayable(&smp_client_data.work_delay);
+	}
+}
+
+static struct smp_client_cmd_req *smp_client_response_discover(const struct smp_hdr *res_hdr)
+{
+	struct smp_hdr smp_header;
+	enum mcumgr_op_t response;
+	struct smp_client_cmd_req *cmd_req;
+
+	if (sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		return NULL;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&smp_client_data.cmd_list, cmd_req, node) {
+		smp_read_hdr(cmd_req->nb, &smp_header);
+		if (smp_header.nh_op == MGMT_OP_READ) {
+			response = MGMT_OP_READ_RSP;
+		} else {
+			response = MGMT_OP_WRITE_RSP;
+		}
+
+		if (smp_header.nh_seq != res_hdr->nh_seq) {
+			continue;
+		} else if (res_hdr->nh_op != response) {
+			continue;
+		}
+
+		return cmd_req;
+	}
+	return NULL;
+}
+
+int smp_client_single_response(struct net_buf *nb, const struct smp_hdr *res_hdr)
+{
+	struct smp_client_cmd_req *cmd_req;
+	smp_client_res_fn cb;
+	void *user_data;
+
+	/* discover request for coming response*/
+	cmd_req = smp_client_response_discover(res_hdr);
+	LOG_DBG("Response Header len %d, flags %d OP: %d group %d id %d seq %d", res_hdr->nh_len,
+		res_hdr->nh_flags, res_hdr->nh_op, res_hdr->nh_group, res_hdr->nh_id,
+		res_hdr->nh_seq);
+
+	if (cmd_req) {
+		cb = cmd_req->cb;
+		user_data = cmd_req->user_data;
+		smp_client_cmd_req_free(cmd_req);
+		if (cb) {
+			cb(nb, user_data);
+		}
+	}
+
+	return MGMT_ERR_EOK;
+}
+
+struct net_buf *smp_client_buf_allocation(uint16_t group, uint8_t command_id, uint8_t op)
+{
+	struct net_buf *nb;
+	struct smp_hdr smp_header;
+
+	nb = smp_packet_alloc();
+
+	if (nb) {
+		/* Write SMP header with payload length 0 */
+		smp_header_init(&smp_header, group, command_id, op, 0);
+		memcpy(nb->data, &smp_header, sizeof(smp_header));
+		nb->len = sizeof(smp_header);
+	}
+	return nb;
+}
+
+void smp_client_buf_free(struct net_buf *nb)
+{
+	while (nb->ref) {
+		smp_packet_free(nb);
+	}
```

Comment:
While? Surely the number of refs should be known and counted properly so it only needs to be free'd once for each time it has been ref'd?

Reply:
Removed while.